Added Life Paths Campaign Options Tab, Added Education Module

IllianiCBT opened the pull request:
# Education Module
## Forward
This PR introduces the "Education Module," a new addition to mhq focused on the education and training of personnel. This functionality is essential for the full implementation of Training Awards into autoAwards.

Within this module, users can send their personnel to local, clan, or prestigious academies (those mentioned in canon sources). Personnel attending these academies will receive bonus XP and Skill Levels. The module fully simulates Clan Sibkos, various graduation states, and events, including the possibility of needing to resit classes, graduating with honors, or washing out of a Caste.

Travel to and from campuses has been abstracted but remains included, as has the consideration of campus destruction or closure.

The module caters primarily to our MechSims user base, although there is potential for non-MechSims users to benefit also.

Overall, this PR adds depth and realism to the game's education system, enhancing the experience for MechSims users while offering some functionality to others.

### Notes
**Prestigious Academies:** Around 30 Prestigious Academies are included in this version. At a later date an additional 120+ will be added in a separate PR.
**autoAwards Integration:** autoAwards functionality has not been added to the Education Module, due to autoAwards' WIP state, but will be added in a separate PR.

## Local Academies
Local Academies refer to institutions without canonical names but are assumed to exist within the game's universe. These include military boot camps, high schools, and boarding schools. These academies are consistently accessible unless the campaign (or personnel) faction is engaged in warfare with all system factions. Warfare data is taken from FactionHints.

## Prestigious Academies
Prestigious Academies are those institutions that have been explicitly mentioned in at least one canon source. The data for these was generous gathered by Hammer. As mentioned above, these are not fully implemented, with only a handful added for testing purposes. Prestigious Academies availability is based on a number of factors, including campaign (or personnel) faction, and location.

## Clan Academies
Clan Academies are only available to campaign factions with the `Clan` tag. There is a good amount of abstraction here, but Clan Creches, Sibkos, and Reeducation Camps are all simulated in both local and non-local forms. This includes an abstracted Trueborn education system and Trials of Position.

## Campaign Options
All module components are fully supported by Campaign Options, allowing users to tweak their experience.

## XML
Based on the awards module code, users can write and edit xml Academy Sets to introduce original local, or prestigious academies. Clan customization is far more limited, but still present. A guide to writing an Academy Set will be included in documentation, once available. Three existing XML have been provided one each for Local, Prestigious, and Clan academy types. All can be disabled through Campaign Options, to discourage users from deleting or editing the default sets (to aid troubleshooting).

## Life Paths
The `Personnel` tab in Campaign Options was getting a little bloated, and this problem is only going to continue as I add more MekSims modules. To solve this issue, I split the `Personnel` tab into `Personnel` and `Life Paths`, moving existing options where appropriate. The idea is that `Life Paths` will be the home of MekSims content, allowing users with no interest in MekSims to effectively ignore those modules. 

## Closes
Closes #3593

IllianiCBT commented on MekHQ/src/mekhq/gui/adapter/PersonnelTableMouseAdapter.java:
```diff
@@ -345,6 +351,24 @@
             }
+            case CMD_BEGIN_EDUCATION: {
+                for (Person person : people) {
+                    EducationController.beginEducation(gui.getCampaign(), person, data[1], data[2], Integer.parseInt(data[3]), data[4], data[5]);
```
In the previous step I converted an Integer into a String, so it could be parsed into `setActionCommand()`, this just reverts that step so should be a non-issue.

IllianiCBT commented on MekHQ/src/mekhq/campaign/personnel/education/Academy.java:
```diff
@@ -0,0 +1,1046 @@
+ */
+@XmlRootElement(name = "academy")
+@XmlAccessorType(value = XmlAccessType.FIELD)
+public class Academy implements Comparable<Academy> {
```
Also a non-issue.

IllianiCBT commented:
Flipped this PR to live, Description has been updated accordingly.

A couple of minor components are missing (as described in Description) and will be added at a later date. My goal for releasing this now, is that I'm aware till Monday, giving people time to review it (should they desire). I can then action those reviewed actions on Monday.

HammerGS commented on MekHQ/data/universe/academies/Prestigious Academies.xml:
```diff
@@ -0,0 +1,940 @@
+        <description>Previously named the Armonstrong Flight Academy, the academy has been relocated and renamed Armstrong Phoenix Academy due to its original campus being rendered uninhabitable by a blood destroying bioweapon. Despite this, it maintains rigorous standards, with intense physical screenings and high skill requirements, leading to over half of the applicants being rejected pre-exam ...
+        <faction>FC</faction>
+        <locationSystem>Delavan</locationSystem>
+        <constructionYear>2070</constructionYear>
```
Should this be 2070?

IllianiCBT replied:
Definitely not, there is also a typo in description. Will fix on Monday

IllianiCBT commented:
I’m aware of the new conflicts. Will fix on Monday

AaronGullickson commented:
I appreciate the refactoring of a LifePaths tab, although part of me thinks why not just call it MekSims 🤣

HammerGS commented:
> I appreciate the refactoring of a LifePaths tab, although part of me thinks why not just call it MekSims 🤣

I prefer Mek Life.

IllianiCBT commented:
Added documentation and addressed the conflicts and other feedback that was provided while I was away. Also caught a couple of minor bugs while writing the documentation, so figured I'd add the documentation to this PR instead of a separate one - seeing that I needed to push a commit anyways.

HammerGS commented:
Picked up conflicts and not sure if this is still a draft

IllianiCBT commented:
> Picked up conflicts and not sure if this is still a draft

Resolved. This PR isn't in draft. Anything remaining will come in a follow-up PR.

rjhancock commented on MekHQ/src/mekhq/campaign/personnel/education/EducationController.java:
```diff
@@ -0,0 +1,1879 @@
+        enrollPerson(campaign, person, academy, campus, faction, courseIndex);
+
+        if (academy.isClan()) {
+            campaign.addReport(resources.getString("offToSchoolClan.text")
+                    .replaceAll("0", person.getFullName())
+                    .replaceAll("1", person.getEduAcademyName())
+                    .replaceAll("2", String.valueOf(person.getEduDaysOfTravelToAcademy())));
+        } else {
+            campaign.addReport(resources.getString("offToSchool.text")
+                    .replaceAll("0", person.getFullName())
+                    .replaceAll("1", person.getEduAcademyName())
+                    .replaceAll("2", String.valueOf(person.getEduDaysOfTravelToAcademy())));
+        }
```
Adjust to use String interpolation instead. You're already making the localized text so will know what goes where and is more secure.

IllianiCBT replied:
Very easy fix. I didn't know about `String.format()` when writing this. I now do, and I absolutely agree that's a far better way of handling these.

It should have occurred to me to go back and use it here, too.

rjhancock commented on MekHQ/src/mekhq/campaign/personnel/education/EducationController.java:
```diff
@@ -0,0 +1,1879 @@
+     * @param resources  The resource bundle containing localized strings.
+     */
+    private static void graduateLaborCaste(Campaign campaign, Person person, Academy academy, ResourceBundle resources) {
+        campaign.addReport(person.getHyperlinkedName() + ' ' + resources.getString("graduatedTrialPassed.text")
+                .replaceAll("0 p", "P")
+                .replaceAll("1", resources.getString("graduatedLabor.text")));
```
String interpolation here s well....

rjhancock commented on MekHQ/src/mekhq/campaign/personnel/education/EducationController.java:
```diff
@@ -0,0 +1,1879 @@
+        int graduationRoll = Compute.randomInt(100);
+
+        if (graduationRoll < 25) {
+            campaign.addReport(person.getHyperlinkedName() + ' ' + resources.getString("graduatedTrialFailed.text")
+                    .replaceAll("0", resources.getString("graduatedTechnician.text")));
```
And here....

You get the idea. not going to mark them all. lol

rjhancock commented on MekHQ/src/mekhq/campaign/personnel/education/EducationController.java:
```diff
@@ -0,0 +1,1879 @@
+     * @param isGraduating A boolean value indicating whether the person is graduating from the academy or not.
+     */
+    private static void improveSkills(Campaign campaign, Person person, Academy academy, Boolean isGraduating) {
+        String[] curriculum = academy.getCurriculums().get(person.getEduCourseIndex())
+                .replaceAll(", ", ",").split(",");
```
Might I suggest doing the split first the streaming the array and striping the text for blank text on either side.

rjhancock commented on MekHQ/src/mekhq/gui/adapter/PersonnelTableMouseAdapter.java:
```diff
@@ -1210,9 +1229,7 @@ protected Optional<JPopupMenu> createPopupMenu() {
         JMenuHelpers.addMenuIfNonEmpty(popup, new AssignPersonToUnitMenu(gui.getCampaign(), selected));
 
         if (oneSelected && person.getStatus().isActive()) {
-            if (gui.getCampaign().getCampaignOptions().isUseManualMarriages()
-                    && (gui.getCampaign().getMarriage().canMarry(gui.getCampaign(),
-                            gui.getCampaign().getLocalDate(), person, false) == null)) {
+            if (gui.getCampaign().getCampaignOptions().isUseManualMarriages() && (gui.getCampaign().getMarriage().canMarry(gui.getCampaign(), gui.getCampaign().getLocalDate(), person, false) == null)) {
```
Needs to be formatted for easier reading.

IllianiCBT replied:
Gah, I must have mucked something up when fixing the bad merge, because I didn't manually make these changes.

I'll make sure to go through the whole file and check to ensure there aren't more mess-ups. Thanks for catching this!

rjhancock commented on MekHQ/src/mekhq/gui/adapter/PersonnelTableMouseAdapter.java:
```diff
@@ -1221,36 +1238,23 @@ protected Optional<JPopupMenu> createPopupMenu() {
                 LocalDate today = gui.getCampaign().getLocalDate();
 
                 // Get all safe potential spouses sorted by age and then by surname
-                final List<Person> personnel = gui.getCampaign().getPersonnel().stream()
-                        .filter(potentialSpouse -> gui.getCampaign().getMarriage().safeSpouse(
-                                gui.getCampaign(), gui.getCampaign().getLocalDate(), person,
-                                potentialSpouse, false))
-                        .sorted(Comparator.comparing((Person p) -> p.getAge(today))
-                                .thenComparing(Person::getSurname))
-                        .collect(Collectors.toList());
+                final List<Person> personnel = gui.getCampaign().getPersonnel().stream().filter(potentialSpouse -> gui.getCampaign().getMarriage().safeSpouse(gui.getCampaign(), gui.getCampaign().getLocalDate(), person, potentialSpouse, false)).sorted(Comparator.comparing((Person p) -> p.getAge(today)).thenComparing(Person::getSurname)).collect(Collectors.toList());
```
Same here and a few other places. It is difficult to read like this.

rjhancock commented on MekHQ/src/mekhq/gui/adapter/PersonnelTableMouseAdapter.java:
```diff
@@ -1221,36 +1238,23 @@ protected Optional<JPopupMenu> createPopupMenu() {
 
                 for (final Person potentialSpouse : personnel) {
                     final String status;
                     final String founder = potentialSpouse.isFounder() ? resources.getString("spouseFounder.text") : "";
                     if (potentialSpouse.getPrisonerStatus().isBondsman()) {
-                        status = String.format(resources.getString("marriageBondsmanDesc.format"),
-                                potentialSpouse.getFullName(), potentialSpouse.getAge(today),
-                                potentialSpouse.getRoleDesc(), founder);
+                        status = String.format(resources.getString("marriageBondsmanDesc.format"), potentialSpouse.getFullName(), potentialSpouse.getAge(today), potentialSpouse.getRoleDesc(), founder);
                     } else if (potentialSpouse.getPrisonerStatus().isCurrentPrisoner()) {
-                        status = String.format(resources.getString("marriagePrisonerDesc.format"),
-                                potentialSpouse.getFullName(), potentialSpouse.getAge(today),
-                                potentialSpouse.getRoleDesc(), founder);
+                        status = String.format(resources.getString("marriagePrisonerDesc.format"), potentialSpouse.getFullName(), potentialSpouse.getAge(today), potentialSpouse.getRoleDesc(), founder);
                     } else {
-                        status = String.format(resources.getString("marriagePartnerDesc.format"),
-                                potentialSpouse.getFullName(), potentialSpouse.getAge(today),
-                                potentialSpouse.getRoleDesc(), founder);
+                        status = String.format(resources.getString("marriagePartnerDesc.format"), potentialSpouse.getFullName(), potentialSpouse.getAge(today), potentialSpouse.getRoleDesc(), founder);
```
And here.... Be mindful of formatters (gets me as well) and make sure the code is more human readable friendly.

rjhancock reviewed (commented):
Sanity looks fine but a few things that I think might need improvement.

Sleet01 reviewed (approved):
Looks good!
